### ISSUE TITLE
Introduce initial caching support

### ISSUE THREAD

**Issue by sdeleuze** [spring-projects-experimental/spring-native#465]
We have basic configuration for caching in `CachingHints`, but EhCache is not supported, and enabling caching support in PetClinic JPA with the following configuration:

```java
@Configuration
@EnableCaching
class CacheConfiguration {

	@Bean
	public JCacheManagerCustomizer petclinicCacheConfigurationCustomizer() {
		return cm -> {
			cm.createCache("vets", cacheConfiguration());
		};
	}

	private javax.cache.configuration.Configuration<Object, Object> cacheConfiguration() {
		return new MutableConfiguration<>().setStatisticsEnabled(true);
	}

}
```

Generates this error that seems to indicate the need for a processor:

```
org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'vetController' defined in class path resource [org/springframework/samples/petclinic/vet/VetController.class]: Unsatisfied dependency expressed through constructor parameter 0; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'vetRepository': Post-processing of FactoryBean's singleton object failed; nested exception is com.oracle.svm.core.jdk.UnsupportedFeatureError: Proxy class defined by interfaces [interface org.springframework.samples.petclinic.vet.VetRepository, interface org.springframework.data.repository.Repository, interface org.springframework.transaction.interceptor.TransactionalProxy, interface org.springframework.aop.framework.Advised, interface org.springframework.core.DecoratingProxy, interface java.io.Serializable] not found. Generating proxy classes at runtime is not supported. Proxy classes need to be defined at image build time by specifying the list of interfaces that they implement. To define proxy classes use -H:DynamicProxyConfigurationFiles=<comma-separated-config-files> and -H:DynamicProxyConfigurationResources=<comma-separated-config-resources> options.
```

We should also document what kind of caching is supported (see the list in `CacheType`) and bring back caching support in the `petclinic-jpa` sample as configured in https://github.com/spring-projects/spring-petclinic.

**Comment by sdeleuze** [comment]
As a follow-up of #481, here is my proposal for a more efficient caching support.

First, once https://github.com/spring-projects/spring-boot/pull/25321 is merged, we should configure `CacheConfigurations` to be initialized at build time to avoid using the cache configuration classes at runtime if they are not useful.

Second, we should update `CacheConfigurationImportSelector` substitution to add classpath check (based on manual conversion of `@ConditionalOnClass` to return only the relevant ones.

Third, we should update `CacheHints` to dynamically add the relevant hints based on a build-time execution of `CacheConfigurationImportSelector`.

**Comment by sdeleuze** [comment]
Tentatively scheduled for 0.9.0 since that's an important feature.

**Comment by snicoll** [comment]
@sdeleuze I can see this is blocked. Anything I can help you with?

**Comment by sdeleuze** [comment]
Thanks for your help offer. I am not blocked anymore by #526, I have updated the branch as discussed to use `JCacheCacheConfiguration` for the `petclinic-jpa` sample but for now `CacheCondition` does not match on native while it does on the JVM.

```
JCacheCacheConfiguration:
      Did not match:
         - Cache org.springframework.boot.autoconfigure.cache.JCacheCacheConfiguration unknown cache type (CacheCondition)
      Matched:
         - @ConditionalOnClass found required classes 'javax.cache.Caching', 'org.springframework.cache.jcache.JCacheCacheManager' (OnClassCondition)
```

I suspect an error on binding but no log since there is an empty `catch (BindException ex) { }`, and that fails even after adding `CacheType` reflection configuration. I am trying to add a substitution to remove that empty catch and get more logs.

**Comment by sdeleuze** [comment]
@aclement In the generated `resource-config.json`, is it expected to have `"pattern": "org/springframework/boot/autoconfigure/cache/CacheAutoConfiguration.CacheConfigurationImportSelector.class"` while the filename is `CacheAutoConfiguration$CacheConfigurationImportSelector.class` ?

**Comment by sdeleuze** [comment]
Thanks for the offer, I am progressing again on it after the discussion we had on #526.

**Comment by sdeleuze** [comment]
I gave it another try with Ehcache 3 and JSR 107, and ended up being blocked on unsafe access in [ThreadLocalRandomUtil](https://github.com/ehcache/ehcache3/blob/v3.9.0/impl/src/unsafe/java/org/ehcache/impl/internal/concurrent/ThreadLocalRandomUtil.java) that I was not able to make it work despite what I think is proper unsafe field configuration. This is likely fixable but will require more time.

That issue + the fact that we need a processor for creating dynamically the proxy configuration + the fact that we need time for proper testing, feedback and adding support for other mechanism clearly indicate to me that this is not doable for 0.9.0.

I have pushed my work on https://github.com/sdeleuze/spring-native/commit/caching. Let's get back on this for 0.10.0.

**Comment by goafabric** [comment]
It would also be nice to have caffeine working.
While SimpleCache works btw. out of the box for now,
Caffeine needs a lot of Typehints in Version 2.8.x 

Following this issue which implicates that Quarkus and Micronaut got it working:
https://github.com/ben-manes/caffeine/issues/434
I put a lot of Typehints (see below)

But still more and more Hints would be needed, because of the excessive use of Reflection.
Version 3.x uses much less Reflection, but there i also gave up after hours.
Would be great to see this supported by Spring

-- cut - not enought to get it working ---

@TypeHint(typeNames = {
        "com.github.benmanes.caffeine.cache.PSW", "com.github.benmanes.caffeine.cache.PSWMS",
        "com.github.benmanes.caffeine.cache.SSLA", "com.github.benmanes.caffeine.cache.SSLMSW",
        "com.github.benmanes.caffeine.cache.SSMSW", "com.github.benmanes.caffeine.cache.SSMSA",
        "com.github.benmanes.caffeine.cache.BLCHeader$DrainStatusRef"
        },
        access = AccessBits.ALL
        )

**Comment by ben-manes** [comment]
Caffeine v3 no longer uses Unsafe, so that might help you get farther if somehow blocked on Ehcache due to that.

The Quarkus folks [suggested](https://github.com/quarkusio/quarkus/issues/10420#issuecomment-654483795) a compile-time evaluation, but I don't really understand Graal enough to implement and verify the suggestion. Any help making this easier for integrators would be appreciated.

**Comment by goafabric** [comment]
> Caffeine v3 no longer uses Unsafe, so that might help you get farther if somehow blocked on Ehcache due to that.
> 
> The Quarkus folks [suggested](https://github.com/quarkusio/quarkus/issues/10420#issuecomment-654483795) a compile-time evaluation, but I don't really understand Graal enough to implement and verify the suggestion. Any help making this easier for integrators would be appreciated.

Thanks for your reply .. actually thats what i tried out, Caffeine v3 because it uses much less reflection ... unfortunately i still could not get this running, because still a lot of hints needed to be added, that i could not figure out.
I might eventually try out v3 again if this seems to be the right choice

**Comment by goafabric** [comment]
@ben-manes oh sorry .. 
I got it wrong .. you are actually the developer of caffeine .. so I guess you know better :)
first and foremost thx for this awesome piece of software ... 
but as stated .. even with unsafe gone, from what I also saw in the code, its not working, at least not of the box,
because i suspect further things need to be added

**Comment by ben-manes** [comment]
Thanks for trying to get it all working. I’m not familiar with Graal and I’d be open to suggestions to make Caffeine better at supporting it. Sorry that you have to fight through this.

https://github.com/oracle/graal/issues/3028 sounds like VarHandles are not supported, leading to improper dead code elimination. I guess it is incompatible with 3.x for now.

In 2.x there is a jandex resource that lists all of the reflections. It was removed in v3 as Quarkus decided it wasn’t necessary, but maybe that could help you bootstrap 2.x. Otherwise I’m at a loss as well.

**Comment by goafabric** [comment]
Thank you for your support.
I think your suggestions should be a great help to ger it working.
I‘ll give it a spin eventually

**Comment by sdeleuze** [comment]
Let's try to provide initial caching support for Redis, the new AOT engine from 0.11 should make it much easier.

**Comment by linghengqian** [comment]
I didn't see any discussion about Spring Native in the Github issue of EhCache. Based on the above discussion, can I assume that Spring Native and JCache (aka JSR107) configured via EhCache are still unavailable?

**Comment by snicoll** [comment]
The issue is still open, so yes. To a large extent, this is a bout adding the necessary hints on third party libraries though so the fix wouldn't probably come here.

**Comment by linghengqian** [comment]
- I handle the Ehcache nativeTest successfully in https://github.com/oracle/graalvm-reachability-metadata/pull/124 .

**Comment by sdeleuze** [comment]
Spring Native is now superseded by Spring Boot 3 official native support, see [the related reference documentation](https://docs.spring.io/spring-boot/docs/current/reference/html/native-image.html) for more details.

As a consequence, I am closing this issue, and recommend trying your use case with latest Spring Boot 3 version. If you still experience the issue reported here, please open an issue directly on the related Spring project (Spring Framework, Data, Security, Boot, Cloud, etc.) with a reproducer.

Thanks for your contribution on the experimental Spring Native project, we hope you will enjoy the official native support introduced by Spring Boot 3.
